[HDF5] Adding flags read/write capabilities

sunethwarna opened the pull request:
This adds flag read write capabilities to nodal and elemental data value containers. Tests are also extended for flags.

sunethwarna committed:
adding element flags

sunethwarna committed:
adding nodal flags

sunethwarna committed:
adding tests

sunethwarna commented:
Is this CI failure in Ubuntu related to this PR? It fails in MeshingApplication

philbucher commented:
the failure is in the CompressiblePotetialApp, unfortunately we don't know why
=> unrelated to this PR

philbucher commented on applications/HDF5Application/custom_io/hdf5_element_data_value_io.cpp:
```diff
@@ -112,6 +145,25 @@ class ReadElementVariableFunctor
+                    std::string const& rVariableName)
+    {
+        Vector<int> data;
+        rFile.ReadDataSet(rPrefix + "/ElementDataValues/" + rVariableName, data,
```
```suggestion
        rFile.ReadDataSet(rPrefix + "/ElementalFlags/" + rVariableName, data,
```

I think the name is misleading

philbucher commented on applications/HDF5Application/custom_io/hdf5_element_data_value_io.cpp:
```diff
@@ -232,13 +287,14 @@ void ElementDataValueIO::ReadElementResults(ElementsContainerType& rElements)
     // Read local data for each variable.
     for (const std::string& r_variable_name : mVariableNames)
-        RegisteredVariableLookup<Variable<array_1d<double, 3>>,
+        RegisteredVariableLookup<Flags,
```
hm this does not seems like it is intended for `Flags`

philbucher reviewed (commented):
generally I don't have objections, but it seems like you are using sth that was designed for variables also for the flags

Can we discuss briefly in person?

sunethwarna committed:
registered_variable_lookup to registered_component_lookup

sunethwarna committed:
adding mpi nodal non-historical data test

sunethwarna committed:
removing slip test

sunethwarna committed:
adding templated container component io

sunethwarna committed:
porting element data value io

sunethwarna committed:
porting nodal data value io

sunethwarna committed:
changing cmake

sunethwarna committed:
adding elemental and nodal flags io

sunethwarna committed:
fixing test bugs

sunethwarna committed:
adding core tests

sunethwarna committed:
adding model part io tests

sunethwarna committed:
adding classes to model part io

sunethwarna committed:
adding to xdmf

sunethwarna committed:
minor

sunethwarna committed:
adding flags to python processes

sunethwarna committed:
adding xdmf testing

sunethwarna committed:
adding hdf5 process tests

sunethwarna committed:
minor

sunethwarna committed:
adding mpi test

sunethwarna committed:
minor

sunethwarna commented:
This is ready to review, I added a new templated class (ContainerComponentIO) which is used to create ElementDataValueIO, ElementFlagValueIO, NodalDataValueIO, NodalFlagValueIO. Some bug fixes in test_hdf5_core.py is also done.

Added all the necessary tests.

sunethwarna commented:
ping @philbucher

sunethwarna committed:
stopping patchers

sunethwarna committed:
remove trailing white spaces

sunethwarna committed:
Merge remote-tracking branch 'origin/master' into hdf5-feature-adding-flags-read-write-output

sunethwarna committed:
adding SynchronizeNodalFlags

sunethwarna committed:
adding condition io

sunethwarna committed:
adding condition io to python core

sunethwarna committed:
ading condition create processes

sunethwarna committed:
adding cpp tests

sunethwarna committed:
adding condition settings to processes

sunethwarna committed:
adding condition tests

sunethwarna committed:
adding condition data

philbucher commented on applications/HDF5Application/CMakeLists.txt:
```diff
@@ -15,11 +15,10 @@ set( KRATOS_HDF5_APPLICATION_SOURCES
 	${CMAKE_CURRENT_SOURCE_DIR}/custom_io/hdf5_data_value_container_io.cpp
```
@sunethwarna I recommend to refactor the CMake in the future to match the template app, this should simplify things

philbucher commented on applications/HDF5Application/python_scripts/single_mesh_xdmf_output_process.py:
```diff
@@ -154,15 +171,31 @@ def CreateCoreSettings(user_settings):
+                                user_settings["element_flag_value_settings"]),
+        CreateOperationSettings("condition_flag_value_output",
+                                user_settings["condition_flag_value_settings"]),
+        CreateOperationSettings("condition_data_value_output",
```
can you please make sure the names match the ones in the VtkOutput?

philbucher commented on applications/HDF5Application/python_scripts/core/xdmf.py:
```diff
@@ -383,6 +383,44 @@ def create_xml_element(self):
         e.append(self.data.create_xml_element())
         return e
 
+class ConditionData(Attribute):
```
The implementation is the same as for `ElementData`? If so then can you
- Use the same? (probably rename to `GeometricalObjectData`
- If not then make a common baseclass?

sunethwarna replied:
done :)

sunethwarna committed:
remove trailing white spaces

philbucher commented on applications/HDF5Application/custom_io/hdf5_container_component_io.cpp:
```diff
@@ -0,0 +1,763 @@
+    {
+        for (auto& p_node : rGhostNodes)
+            p_node->GetValue(rComponent) = rComponent.Zero();
+        rCommunicator.AssembleNonHistoricalData(rComponent);
```
you can sync non-hist variables too

sunethwarna replied:
done :)

philbucher commented on applications/HDF5Application/custom_io/hdf5_condition_data_value_io.h:
```diff
@@ -0,0 +1,115 @@
+
+    void ReadConditionResults(ConditionsContainerType& rConditions)
+    {
+        Communicator dummy_communicator;
```
why don't you pass it from outside like you do for the nodes?
Even though not needed atm it might be useful in the future

sunethwarna replied:
done :)

philbucher commented on applications/HDF5Application/python_scripts/core/operations/model_part.py:
```diff
@@ -105,6 +105,69 @@ def __call__(self, model_part, hdf5_file):
+
+    def __call__(self, model_part, hdf5_file):
+        KratosHDF5.HDF5ConditionFlagValueIO(
+            self.GetSettings(model_part).Get(), hdf5_file).ReadConditionFlags(model_part.Conditions)
```
why not passing the `Communicator` here too? This could unify implementations in the `hdf5_container_component_io`
Same for other cases where you don't pass the `Communicator`

sunethwarna replied:
done :)

philbucher reviewed (commented):
Overall looks good to me :+1:

some questions, see the comments

So now the conditions are always written to the Hdf file? Could this be optional (thinking abt file size). Or only write if you want to plot on it...? Not sure how it works/is intended with #6363

sunethwarna committed:
removing ghost elements

sunethwarna committed:
making GeometricalObjectData base class

sunethwarna committed:
adding communicator

sunethwarna committed:
Merge remote-tracking branch 'origin/master' into hdf5-feature-adding-flags-read-write-output

sunethwarna commented:
@philbucher I made the changes. Could you have a look?

sunethwarna committed:
making public

philbucher reviewed (approved):
:+1:
